### PR TITLE
Fix: 수강 신청 시 발생하는 동시성 문제 해결

### PR DESCRIPTION
**Pull request by yeonise**
## Key Feature

- 강의의 정원을 초과하여 수강 신청이 되는 문제를 해결했습니다.
- EnrollmentService 테스트 코드를 리팩토링했습니다.

## To reviewer

[프로젝트 블로그](https://velog.io/@on-and-off/Redis%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EC%97%AC-%EB%8F%99%EC%8B%9C%EC%84%B1-%EB%AC%B8%EC%A0%9C-%ED%95%B4%EA%B2%B0%ED%95%98%EA%B8%B0-fmo9ufhx)

### CODE REVIEW

**Comment by ghkdgus29** [comment]
### Test Coverage Report
|Overall Project|95.21% **`-2.16%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|36.03%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EnrollmentController.java](https://github.com/mujik-tigers/course-registration-system/blob/df1ed774e625bad7041ebf44b57bcc415785ad74/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Fpresentation%2FEnrollmentController.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/course-registration-system/blob/df1ed774e625bad7041ebf44b57bcc415785ad74/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[EnrollmentService.java](https://github.com/mujik-tigers/course-registration-system/blob/df1ed774e625bad7041ebf44b57bcc415785ad74/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Fapplication%2FEnrollmentService.java)|100%|:green_apple:|
|[DuplicateSubjectException.java](https://github.com/mujik-tigers/course-registration-system/blob/df1ed774e625bad7041ebf44b57bcc415785ad74/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fenrollment%2FDuplicateSubjectException.java)|100%|:green_apple:|
|[RedissonEnrollmentLockFacade.java](https://github.com/mujik-tigers/course-registration-system/blob/df1ed774e625bad7041ebf44b57bcc415785ad74/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Ffacade%2FRedissonEnrollmentLockFacade.java)|17.14% **`-82.86%`**|:x:|

File: .github/workflows/auto-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+name: Auto Deploy to EC2
```
Comment (ghkdgus29) [inline .github/workflows/auto-deploy.yml]: ~자동 배포파일 제거해주셔야 될듯요~!!~

File: src/main/java/site/courseregistrationsystem/enrollment/application/EnrollmentService.java [hunk 2]
```diff
@@ -149,7 +127,7 @@ public EnrolledLectures fetchAll(Long studentPk) {
 	}
 
 	public EnrollmentCapacity fetchCountBy(Year year, Semester semester, Long lectureId) {
-		Lecture lecture = lectureRepository.findWithSchedule(lectureId).orElseThrow(LectureNotFoundException::new);
+		Lecture lecture = lectureRepository.findByIdWithSubjectAndSchedule(lectureId).orElseThrow(LectureNotFoundException::new);
```
Comment (ghkdgus29) [inline src/main/java/site/courseregistrationsystem/enrollment/application/EnrollmentService.java:130]: fetch join 으로 subject 와 schedule 을 전부 가져오기에 이렇게 명명하신거죠?!

File: src/main/java/site/courseregistrationsystem/enrollment/facade/RedissonEnrollmentLockFacade.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		Lecture lecture = lectureRepository.findByNumberWithSubjectAndSchedule(lectureNumber).orElseThrow(LectureNotFoundException::new);
+
+		return enrollWithLock(now, student, lecture);
+	}
+
+	private EnrolledLecture enrollWithLock(LocalDateTime now, Student student, Lecture lecture) {
+		RLock lock = redissonClient.getLock(lecture.getId().toString());
+
+		try {
+			boolean available = lock.tryLock(lockWaitTime, lockLeaseTime, lockTimeUnit);
```
Comment (ghkdgus29) [inline src/main/java/site/courseregistrationsystem/enrollment/facade/RedissonEnrollmentLockFacade.java:52]: 이 tryLock 의 의미하는 것이 무엇인지 설명해주실수 있나요? (궁금)

**Comment by ghkdgus29** [comment]
수고하셨습니다~!!

동시성 문제를 멋지게 해결하셨네요~~

동시성 문제를 테스트하는 방법은 저도 한번 연구해봐야겠네요

~auto_deploy.yml 파일만 제거해서 push 하시면 머지해드릴게용~

File: .github/workflows/auto-deploy.yml (hunk 1 above)
Comment (ghkdgus29) [reply .github/workflows/auto-deploy.yml]: 이 부분 제가 제거할게요 ~

Comment (yeonise) [reply .github/workflows/auto-deploy.yml]: 감사링~~~~~ 😎

File: src/main/java/site/courseregistrationsystem/enrollment/application/EnrollmentService.java (hunk 2 above)
Comment (yeonise) [reply src/main/java/site/courseregistrationsystem/enrollment/application/EnrollmentService.java:130]: 네 고렇습니다!?

File: src/main/java/site/courseregistrationsystem/enrollment/facade/RedissonEnrollmentLockFacade.java (hunk 3 above)
Comment (yeonise) [reply src/main/java/site/courseregistrationsystem/enrollment/facade/RedissonEnrollmentLockFacade.java:52]: `lockWaitTime` 시간 동안 Lock 획득을 시도하고, Lock 획득에 성공한다면 최대 `lockLeaseTime` 시간동안 Lock을 점유하고 있을 수 있습니다! 만약 `lockWatiTime` 동안 Lock 획득에 실패하게 된다면 `boolean available` 은 `false` 가 되겠네욤!

[참고한 블로그](https://incheol-jung.gitbook.io/docs/q-and-a/spring/redisson-trylock)

**Comment by ghkdgus29** [comment]
### Test Coverage Report
|Overall Project|97.09% **`-0.27%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|91.91%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EnrollmentController.java](https://github.com/mujik-tigers/course-registration-system/blob/f5371a76f34f73bbb2c603458ed8e5bb1d4b0ead/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Fpresentation%2FEnrollmentController.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/course-registration-system/blob/f5371a76f34f73bbb2c603458ed8e5bb1d4b0ead/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[EnrollmentService.java](https://github.com/mujik-tigers/course-registration-system/blob/f5371a76f34f73bbb2c603458ed8e5bb1d4b0ead/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Fapplication%2FEnrollmentService.java)|100%|:green_apple:|
|[DuplicateSubjectException.java](https://github.com/mujik-tigers/course-registration-system/blob/f5371a76f34f73bbb2c603458ed8e5bb1d4b0ead/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fenrollment%2FDuplicateSubjectException.java)|100%|:green_apple:|
|[RedissonEnrollmentLockFacade.java](https://github.com/mujik-tigers/course-registration-system/blob/f5371a76f34f73bbb2c603458ed8e5bb1d4b0ead/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fenrollment%2Ffacade%2FRedissonEnrollmentLockFacade.java)|89.52% **`-10.48%`**|:green_apple:|

**Pull request merged**